### ISSUE TITLE
Elastic Indexed Job

### ISSUE THREAD

**Issue by ahg-g** [kubernetes/kubernetes#114862]
### What would you like to be added?

Currently we validate that `spec.completions` is set when `completionMode=Indexed` for job. I suggest we relax this validation. 

This will require the following changes:

1) We need to update the job controller to look at `spec.parallelism` instead of `spec.completions` when calculating the range of indexes.
2) When reducing parallelism, we should remove the higher indexes first (otherwise the controller will eventually do that anyways to ensure that the pods have a continuous range).

both changes only apply when `spec.completions=nil`

### Why is this needed?

The use case is jobs with workers that require a form of autoscaling, like Ray, Spark or Horovord. By modeling these as jobs with `spec.completion=nil`, `spec.parallelism` can be used to scale the job.

While such jobs could be modeled as a StatefulSet, the job API offers batch-specific features that makes it a better fit (like [retriable and non retriable failures](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures))

/sig apps
@soltysh @alculquicondor wdyt?

**Comment by alculquicondor** [comment]
+1
The only thing to note is that if we do the controller changes in release X, then the API validation changes need to go in release X+1

**Comment by sftim** [comment]
If we make this change, what are the criteria for considering the Job as complete?

**Comment by alculquicondor** [comment]
Probably worth bringing this up to the next SIG meeting, to make sure we are not missing anything.

> If we make this change, what are the criteria for considering the Job as complete?

We should use the same behavior that NonIndexed jobs have: once one pod finishes successfully, no new pods are created.

**Comment by sftim** [comment]
:thought_balloon: How about using `Indexed` to describe the completion mode we already know, and `IndexedFirstToComplete` for the new behavior? Something like that.

**Comment by alculquicondor** [comment]
We are not changing existing behavior, because `completions=nil` is just disallowed during validation. Adding a new completionMode sounds like an overkill.

**Comment by sftim** [comment]
My worry: someone might watch for a Job and not write code to check `completions`, because they see that the mode is `"Indexed"`. Their scheduler incorrectly ignores the Job because they think `completions: null` is guaranteed by the API, based on their reading of the docs (even if we never made such a guarantee in terms of API versioning).

**Comment by sftim** [comment]
What can help: draft the change to https://kubernetes.io/docs/concepts/workloads/controllers/job/#completion-mode _first_, and then once we're happy that it's feasible to explain, start on the code.

**Comment by ahg-g** [comment]
> My worry: someone might watch for a Job and not write code to check completions, because they see that the mode is "Indexed". Their scheduler incorrectly ignores the Job because they think completions: null is guaranteed by the API, based on their reading of the docs (even if we never made such a guarantee in terms of API versioning).

Thanks @sftim for the feedback, relaxing validation in general is allowed. As to this specific example, the change we are proposing doesn't break this hypothetical scheduler, it will continue to operate as expected from that scheduler's perspective.

> What can help: draft the change to https://kubernetes.io/docs/concepts/workloads/controllers/job/#completion-mode first, and then once we're happy that it's feasible to explain, start on the code.

 `completions=nil` is historically a permitted mode of operation and have clear semantics, and this proposed change doesn't change those semantics.

**Comment by ahg-g** [comment]
But let me draft one on this issue, I will post in a moment

**Comment by ahg-g** [comment]
### Completion mode

Jobs can have the following modes of operation specified in `.spec.completionMode`:

- `NonIndexed` (default): the Pods of a Job are considered homologous of each other. When `spec.completions` is set, the Job is considered complete when there have been `.spec.completions` successfully completed Pods.
- `Indexed`: the Pods of a Job get an associated completion index from 0 to
  `$(count)-1`, where count is .spec.completions if set, .spec.parallelism if not. The index is available 
     through three mechanisms:
  - The Pod annotation `batch.kubernetes.io/job-completion-index`.
  - As part of the Pod hostname, following the pattern `$(job-name)-$(index)`.
    When you use an Indexed Job in combination with a
    {{< glossary_tooltip term_id="Service" >}}, Pods within the Job can use
    the deterministic hostnames to address each other via DNS. For more information about
    how to configure this, see [Job with Pod-to-Pod Communication](/docs/tasks/job/job-with-pod-to-pod-communication/).
  - From the containerized task, in the environment variable `JOB_COMPLETION_INDEX`.
  - When scaling `spec.parallelism` down, the higher active indexes will be removed first.
  
  When spec.completions is set, and indexed Job is considered complete when there is one successfully completed Pod
  for each index. For more information about how to use this mode, see
  [Indexed Job for Parallel Processing with Static Work Assignment](/docs/tasks/job/indexed-parallel-processing-static/).
  Note that, although rare, more than one Pod could be started for the same
  index, but only one of them will count towards the completion count.

Finally, in both modes, when `spec.completions` is not set, then the job is considered complete when at least one pod successfully completes.

**Comment by sftim** [comment]
I think we need to explain that if `completions` isn't set, indexes can be reused. At least, that's what the phrase
> an associated completion index from 0 to `$(count)-1`, where count is `.spec.completions` if set, `.spec.parallelism` if not

implies.

If we didn't want to reuse the indices, then let's explain (and, by implication, consider the design details for) what happens if we create more Pods than we have parallelism.

I'd like this to have KEP: having an enhancement proposal makes it easier for SIG Release to track the work around updating the docs.

**Comment by ahg-g** [comment]
What do you mean by indexes can be reused?

**Comment by sftim** [comment]
In the Kubernetes blog, [Introducing Indexed Jobs](https://kubernetes.io/blog/2021/04/19/introducing-indexed-jobs/) suggests that
> each worker Pod can have a statically assigned partition of the data based on the index

and someone may have made a container image that relies on that assumption. Does this change create the risk of confusing people who do assume this?

Let's say I have a work queue that will turn out to hold 99 items, and I want to use a parallelism of 10. I don't know the size of the work queue at the time I define the Job. I set this new Job to be indexed because all my other (fixed-size) Jobs are indexed.

Some people will want to have an index that matches the ordinal of running Pods like with a StatefulSet, but other people might use the index for something like, say, interpolating into the filename of the Job's output.

If we can handle both of the cases I've sketched here, great. If not, let's document which case we're supporting and add a [callout](https://kubernetes.io/docs/contribute/style/style-guide/#caution) to alert the people who were expecting the opposite behavior.

(This is why I think it's good to draft the docs first - if we're making a design that requires warning users, we might want to reconsider it).

**Comment by ahg-g** [comment]
> and someone may have made a container image that relies on that assumption. Does this change create the risk of confusing people who do assume this?

Note that he proposed change does not impact existing workloads. We are adding new semantics, not changing existing ones.

> Some people will want to have an index that matches the ordinal of running Pods like with a StatefulSet, but other people might use the index for something like, say, interpolating into the filename of the Job's output.

The proposed semantics supports a StatefulSet like behavior, the indexes will cover the number of replicas (parallelism). The second use case is an interesting one, but it will not be enabled by the proposed change, and I don't know why one would think that if we are clearly stating that the indexes will range from 0 to spec.parallelism-1 and that success is defined as any pod finishing successfully. I will send out a KEP.

> but other people might use the index for something like, say, interpolating into the filename of the Job's output.

I haven't seen a batch API with these semantics, and to address this use case (number of indexes is not defined before hand) the queue can associate the item with an index, not the job. The work queue pattern by design aims to reuse the pods and treat them as stateless because the state is in the queue. 

Slightly related, we are thinking about adding a `completionPolicy` API to Job (e.g., the job should be declared successful if 80% of the pods successfully completes);  so, if we set `spec.completions` to some maximum number and come up with a completion policy that says "exit if the queue is empty", then we could support the case you mention via job, but I am not really convinced that this is the right approach because I believe indexed job isn't the right mode for this use case as I mentioned in the paragraph above.

**Comment by sathyanarays** [comment]
> Finally, in both modes, when spec.completions is not set, then the job is considered successful when at least one pod successfully completes.

Is it more accurate to say "Finally, in both modes, when spec.completions is not set, then the job is considered successful when at least `N` pod successfully completes where N is the `spec.parallelism`"?

**Comment by ahg-g** [comment]
> Is it more accurate to say "Finally, in both modes, when spec.completions is not set, then the job is considered successful when at least N pod successfully completes where N is the spec.parallelism"?

No, that is not the current semantics of `spec.completions=nil`; as explained in  https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs: 
```
once any Pod has exited with success, no other Pod should still be doing any work for this task or writing any output. They should all be in the process of exiting.
```

The behavior you are asking about can be achieved by setting `spec.completions=spec.parallelism`

**Comment by sathyanarays** [comment]
@ahg-g ,
I created a job with the following spec:

```
apiVersion: batch/v1
kind: Job
metadata:
  name: backoff-exp
spec:
  backoffLimit: 4
  parallelism: 10
  template:
    spec:
      containers:
        - name: die
          image: myubuntu:1
          command: ["/bin/sh","-c"]
          args: ["/bin/sleep `rand -M 300` && /bin/true"]
          imagePullPolicy: Never
      restartPolicy: Never
```

Each pod sleeps `random` seconds and exits successfully. In this case, the job was marked completed only after all the 10 pods finished.

#### kubectl get pods
```
NAME                READY   STATUS      RESTARTS   AGE
backoff-exp-5mb6m   0/1     Completed   0          4m41s
backoff-exp-6j76z   0/1     Completed   0          4m41s
backoff-exp-6vv88   0/1     Completed   0          4m41s
backoff-exp-7vp8c   0/1     Completed   0          4m41s
backoff-exp-gtz7n   0/1     Completed   0          4m41s
backoff-exp-q9kvt   0/1     Completed   0          4m41s
backoff-exp-vnw7m   0/1     Completed   0          4m41s
backoff-exp-wngkf   0/1     Completed   0          4m41s
backoff-exp-xhq9x   0/1     Completed   0          4m41s
backoff-exp-z4cc9   0/1     Completed   0          4m41s
``` 

#### Start and Finish time for all the above pods
```
2023-01-09T03:56:35Z | 2023-01-09T03:59:55Z
2023-01-09T03:56:35Z | 2023-01-09T03:59:55Z
2023-01-09T03:56:36Z | 2023-01-09T03:59:56Z
2023-01-09T03:56:36Z | 2023-01-09T04:01:14Z
2023-01-09T03:56:36Z | 2023-01-09T04:01:14Z
2023-01-09T03:56:36Z | 2023-01-09T04:01:14Z
2023-01-09T03:56:36Z | 2023-01-09T04:01:14Z
2023-01-09T03:56:37Z | 2023-01-09T03:59:44Z
2023-01-09T03:56:37Z | 2023-01-09T03:59:44Z
2023-01-09T03:56:37Z | 2023-01-09T03:59:44Z
```

#### Status of the completed job
```
  status:
    completionTime: "2023-01-09T04:01:17Z"
    conditions:
    - lastProbeTime: "2023-01-09T04:01:17Z"
      lastTransitionTime: "2023-01-09T04:01:17Z"
      status: "True"
      type: Complete
    ready: 0
    startTime: "2023-01-09T03:56:34Z"
    succeeded: 10
    uncountedTerminatedPods: {}
```
The job is marked complete only after the last pod has finished. Is the above behavior correct? I think it's not conforming to the line `For a non-parallel Job, you can leave both .spec.completions and .spec.parallelism unset. When both are unset, both are defaulted to 1.` in the [doc](https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs)

Please let me know your thoughts on this!

**Comment by ahg-g** [comment]
Any active pods will not be deleted and will be allowed to run to completion (fail or success), but the job will be marked as **successful** if one pod finishes successfully.

**Comment by ahg-g** [comment]
I have this on today's (Jan 9th) sig-apps meeting

**Comment by ahg-g** [comment]
Meeting outcome: we will write a KEP, but there seems to be an agreement on the general direction as proposed in the initial post.

**Comment by asm582** [comment]
Curious to know If `spec.completions=nil` then will a failed pod will ever be retried?

**Comment by danielvegamyhre** [comment]
PR for Elastic Index Job docs (in progress): https://github.com/kubernetes/website/pull/39999

**Comment by vsoch** [comment]
hey folks! I wanted to check in on the state of this - it looks like the elastic index job docs PR was merged https://github.com/kubernetes/website/pull/39999 what comes next?

**Comment by alculquicondor** [comment]
Since this was completed:
/close

What would you like to see next regarding this topic? However, we probably should open new issue(s)

**Comment by vsoch** [comment]
> Since this was completed:

I thought that might be the case!

> What would you like to see next regarding this topic? However, we probably should open new issue(s)

We are using this for experiments to scale / reduce our Flux MiniClusters and it's working great! I don't have further comments here but will definitely open new issues or bring up discussion as needed.
